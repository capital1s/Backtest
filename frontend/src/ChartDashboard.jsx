import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { Line } from "react-chartjs-2";
import { API_BASE_URL } from "./apiConfig";

const FREQUENCIES = [
  { value: "1min", label: "1 Minute" },
  { value: "5min", label: "5 Minutes" },
  { value: "15min", label: "15 Minutes" },
  { value: "1H", label: "1 Hour" },
];

function ChartDashboard({ ticker, setTicker, tickers }) {
  const [frequency, setFrequency] = useState("1min");
  const [chartData, setChartData] = useState([]);
  const [error, setError] = useState("");
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    async function fetchChart() {
      setLoading(true);
      setError("");
      if (!ticker) {
        setError("No ticker selected. Please choose a ticker.");
        setLoading(false);
        return;
      }
      const allowedFrequencies = ["1min", "5min", "15min", "1H"];
      if (!allowedFrequencies.includes(frequency)) {
        setError("Invalid frequency selected.");
        setLoading(false);
        return;
      }
      try {
        const res = await fetch(`${API_BASE_URL}/minute_chart`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({
            ticker,
            frequency,
            duration: "1 D",
            bar_size: "1 min",
          }),
        });
        if (!res.ok) {
          const text = await res.text();
          setError(`API Error: ${res.status} - ${text}`);
          setLoading(false);
          return;
        }
        const data = await res.json();
        if (!data.chart || !Array.isArray(data.chart)) {
          setError("Malformed API response: missing chart data.");
          setChartData([]);
          setLoading(false);
          return;
        }
        setChartData(data.chart);
        setError("");
      } catch (err) {
        setError(`Network or JSON error: ${err.message || err}`);
        setChartData([]);
      }
      setLoading(false);
    }
    fetchChart();
  }, [ticker, frequency]);

  const data = {
    labels: chartData.map((row) => row.date),
    datasets: [
      {
        label: "Close",
        data: chartData.map((row) => row.close),
        borderColor: "blue",
        fill: false,
      },
    ],
  };

  return (
    <div>
      {loading && <div>Loading chart data...</div>}
      {error && (
        <div style={{ color: "red", marginBottom: "1rem" }} role="alert">
          {error}
        </div>
      )}
      <label>
        Ticker:
        <select value={ticker} onChange={(e) => setTicker(e.target.value)}>
          {tickers.map((t) => (
            <option key={t} value={t}>
              {t}
            </option>
          ))}
        </select>
      </label>
      <label>
        Frequency:
        <select
          value={frequency}
          onChange={(e) => setFrequency(e.target.value)}
        >
          {FREQUENCIES.map((f) => (
            <option key={f.value} value={f.value}>
              {f.label}
            </option>
          ))}
        </select>
      </label>
      {!loading && !error && (
        <Line data={data} options={{}} data-testid="historical-chart-canvas" />
      )}
    </div>
  );
}

ChartDashboard.propTypes = {
  ticker: PropTypes.string.isRequired,
  setTicker: PropTypes.func.isRequired,
  tickers: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default React.memo(ChartDashboard);
