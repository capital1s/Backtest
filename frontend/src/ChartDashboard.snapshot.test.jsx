// @vitest-environment jsdom
import React from "react";
import { describe, it, expect, vi } from "vitest";
import { beforeAll, afterAll } from "vitest";
import { render, act } from "@testing-library/react";
import ChartDashboard from "./ChartDashboard";

// Prevent network calls during tests
beforeAll(() => {
  global.fetch = vi.fn(() => Promise.resolve({
    ok: true,
    json: () => Promise.resolve({ chart: [] }),
    text: () => Promise.resolve("")
  }));
});

afterAll(() => {
  global.fetch && (global.fetch = undefined);
});

describe("ChartDashboard", () => {
  it("matches snapshot", async () => {
    let asFragment;
    await act(async () => {
      const renderResult = render(
        <ChartDashboard ticker="AAPL" setTicker={() => {}} tickers={["AAPL", "MSFT"]} />
      );
      asFragment = renderResult.asFragment;
      // Wait for any async state updates
      await new Promise(resolve => setTimeout(resolve, 0));
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
