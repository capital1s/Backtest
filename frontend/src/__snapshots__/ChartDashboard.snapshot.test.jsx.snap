// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChartDashboard > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      Loading chart data...
    </div>
    <label>
      Ticker:
      <select>
        <option
          value="AAPL"
        >
          AAPL
        </option>
        <option
          value="MSFT"
        >
          MSFT
        </option>
      </select>
    </label>
    <label>
      Frequency:
      <select>
        <option
          value="1min"
        >
          1 Minute
        </option>
        <option
          value="5min"
        >
          5 Minutes
        </option>
        <option
          value="15min"
        >
          15 Minutes
        </option>
        <option
          value="1H"
        >
          1 Hour
        </option>
      </select>
    </label>
  </div>
</DocumentFragment>
`;
